### ISSUE TITLE
`String#dump` returns incorrect character escaping for unicode codepoints > `0xffff`

### ISSUE THREAD

**Issue by postmodern** [oracle/truffleruby#2794]
I noticed a slight difference in how `String#dump` escapes unicode characters greater than `0xffff` on TruffleRuby 22.3.0 vs. CRuby 3.0.4. TruffleRuby should escape unicode characters greater than `0xffff` as `\u{xxxx}` to match CRuby; no clue why CRuby does this.

### TruffleRuby 22.3.0

```ruby
0xffff.chr(Encoding::UTF_8).dump
# =>  "\"\\uFFFF\""
0x10000.chr(Encoding::UTF_8).dump
# => "\"\\u10000\""
```

### CRuby 3.0.4

```ruby
0xffff.chr(Encoding::UTF_8).dump
# => "\"\\uFFFF\""
0x10000.chr(Encoding::UTF_8).dump
# => "\"\\u{10000}\""
```

**Comment by eregon** [comment]
Thanks for the report.
I guess it's done so to know how many digits to look at, i.e., always 4 without `{}`, and what's between `\u{` and `}` otherwise.

**Comment by andrykonchin** [comment]
Fixed in 57e53f8adf99402776f62633873a366e38ea3d61
